src/mem/impls.rs

```diff
@@ -1,27 +1,221 @@
+
+const WORD_SIZE: usize = core::mem::size_of::<usize>();
+const WORD_MASK: usize = WORD_SIZE - 1;
+
+const WORD_COPY_THRESHOLD: usize = if 2 * WORD_SIZE > 16 {
+    2 * WORD_SIZE
+} else {
+    16
+};
+
+#[inline(always)]
+unsafe fn copy_forward_bytes(mut dest: *mut u8, mut src: *const u8, n: usize) {
+    let dest_end = dest.add(n);
+    while dest < dest_end {
+        *dest = *src;
+        dest = dest.add(1);
+        src = src.add(1);
+    }
+}
+
 #[inline(always)]
-pub unsafe fn copy_forward(dest: *mut u8, src: *const u8, n: usize) {
-    let mut i = 0;
-    while i < n {
-        *dest.add(i) = *src.add(i);
-        i += 1;
+unsafe fn copy_forward_aligned_words(dest: *mut u8, src: *const u8, n: usize) {
+    let mut dest_usize = dest as *mut usize;
+    let mut src_usize = src as *mut usize;
+    let dest_end = dest.add(n) as *mut usize;
+
+    while dest_usize < dest_end {
+        *dest_usize = *src_usize;
+        dest_usize = dest_usize.add(1);
+        src_usize = src_usize.add(1);
     }
 }
 
 #[inline(always)]
-pub unsafe fn copy_backward(dest: *mut u8, src: *const u8, n: usize) {
-    // copy from end
-    let mut i = n;
-    while i != 0 {
-        i -= 1;
-        *dest.add(i) = *src.add(i);
+unsafe fn copy_forward_misaligned_words(dest: *mut u8, src: *const u8, n: usize) {
+    let mut dest_usize = dest as *mut usize;
+    let dest_end = dest.add(n) as *mut usize;
+
+    // Calculate the misalignment offset and shift needed to reassemble value.
+    let offset = src as usize & WORD_MASK;
+    let shift = offset * 8;
+
+    // Realign src
+    let mut src_aligned = (src as usize & !WORD_MASK) as *mut usize;
+    // XXX: Could this possibly be UB?
+    let mut prev_word = *src_aligned;
+
+    while dest_usize < dest_end {
+        src_aligned = src_aligned.add(1);
+        let cur_word = *src_aligned;
+        #[cfg(target_endian = "little")]
+        let resembled = prev_word >> shift | cur_word << (WORD_SIZE * 8 - shift);
+        #[cfg(target_endian = "big")]
+        let resembled = prev_word << shift | cur_word >> (WORD_SIZE * 8 - shift);
+        prev_word = cur_word;
+
+        *dest_usize = resembled;
+        dest_usize = dest_usize.add(1);
+    }
+}
+
+#[inline(always)]
+pub unsafe fn copy_forward(mut dest: *mut u8, mut src: *const u8, mut n: usize) {
+    if likely(n >= WORD_COPY_THRESHOLD) {
+        // Align dest
+        // Because of n >= 2 * WORD_SIZE, dst_misalignment < n
+        let dest_misalignment = (dest as usize).wrapping_neg() & WORD_MASK;
+        copy_forward_bytes(dest, src, dest_misalignment);
+        dest = dest.add(dest_misalignment);
+        src = src.add(dest_misalignment);
+        n -= dest_misalignment;
+
+        let n_words = n & !WORD_MASK;
+        let src_misalignment = src as usize & WORD_MASK;
+        if likely(src_misalignment == 0) {
+            copy_forward_aligned_words(dest, src, n_words);
+        } else {
+            copy_forward_misaligned_words(dest, src, n_words);
```

Comment:
Well, first of all `ptr::read_unaligned` and `ptr::write_aligned` calls `ptr::copy_nonoverlapping` which translates to memcpy, so I would like to avoid the possibility of an recursion if LLVM doesn't optimise them away.

Secondly, `ptr::read_unaligned` has really poor performance. On platforms without misaligned memory access support it will translate to `size_of<usize>()` number of byte loads.

This branch is necessary because we don't want to bear the burden of all the shifts and checks necessary for misaligned loads if dest and src are perfectly co-aligned.

Reply:
Have you tried this out and seen infinite recursion? Have you tried it out and seen if it's slower?

Reply:
There'll be an infinite recursion if compiled in debug mode.

On architectures that does not support misaligned loads (so most ISAs other than armv8 and x86/64) the performance is much slower because it generates 8 byte load and 16 bit ops rather than 1 word load and 3 bit ops.